lib/jumubase_web/views/encoders/ecto_encoders.ex

```diff
@@ -1,18 +1,29 @@
-# The following Poison.Encoder implementations ensure that JSON-encoding an
+# The following Jason.Encoder implementations ensure that JSON-encoding an
 # Ecto changeset or record doesn't result in errors because of un-encodable
 # constructs (such as keyword lists).
 
-defimpl Poison.Encoder, for: Ecto.Changeset do
+require Protocol
+alias Jumubase.Showtime.{Appearance, Participant, Performance, Piece}
+
+# White-list schema fields that may be exposed via JSON
+Protocol.derive(Jason.Encoder, Performance, only: [:id, :contest_category_id, :appearances, :pieces])
+Protocol.derive(Jason.Encoder, Appearance, only: [:id, :instrument, :role, :participant])
+Protocol.derive(Jason.Encoder, Participant, only: [:id, :given_name, :family_name, :birthdate, :phone, :email])
+Protocol.derive(Jason.Encoder, Piece, only: [:id, :title, :composer, :composer_born, :composer_died, :artist, :epoch, :minutes, :seconds])
+
+defimpl Jason.Encoder, for: Ecto.Changeset do
   import JumubaseWeb.ErrorHelpers, only: [translate_error: 1]
 
-  def encode(changeset, options) do
+  def encode(changeset, _options) do
+    data = if changeset.action == :insert, do: %{}, else: changeset.data
+
     output = %{
-      data: changeset.data,
+      data: data,
```

Comment:
Afaics, this is the place where we need to apply a custom recursive filter which

* should remove `Ecto.Association.NotLoaded` fields (or at least set them to `nil`)
* should remove `__struct__` and `__meta__` fields
* should remove "private" fields that aren't needed to populate the registration form

Unfortunately, I couldn't figure out how to write such a filter since Elixir structs don't implement the `Enumerable` protocol 😕